---
title: "AnthoMicroComp"
author: "Lauren Mee"
date: "2022-10-26"
output: html_document
---

26th October 2022

```{r}
set.seed(1517)
dir.create("output/")
```

## Libraries

```{r}
library(ggplot2)
library(dplyr)
library(stringr)
library(reshape)
library(phyloseq)
```

## Resources

Tutorial discussing different methods and approaches to assesing microbiome data from metagenomic sources:
https://yanhui09.github.io/microbiome_analysis/1_microbiome_r.html#what-you-will-work-on
(includes DESeq2 approaches and a good heatmap alternative that looks interesting)

Microbial Community Diversity Analysis Tutorial with Phyloseq:
http://deneflab.github.io/MicrobeMiseq/demos/mothur_2_phyloseq.html
(may not be as useful, looks to be explicitly for amplicon data)

Introduction to the Statistical Analysis of Microbiome Data in R
https://www.nicholas-ollberding.com/post/introduction-to-the-statistical-analysis-of-microbiome-data-in-r/

Tutorial for making phyloseq objects
https://joey711.github.io/phyloseq/import-data.html

"Breakaway" Approach to estimating richness
https://adw96.github.io/breakaway/articles/intro-diversity-estimation.html

## Experiment Notes

The purpose of this project is to data mine existing publc RNAseq and other resources and use the CZID.org tool in order to elucidate potential patterns of microbial content across the sample species used therein. 

Samples were chosen from SRA projects that used Hymenopteran species and also met the following criteria: 1) the tissue used was or included the abdomen, 2) nothing experimental had happened to the individual sample that would be an obvious influence upon the microbiome, i.e. no infection administered / particularly stressed individuals and 3) sociality of the sample was known. 

These chosen samples were then processed as follows:

- the phylogenetically nearest available genome was downloaded from NCBI per sample species. This was then indexed using STAR

- the sample was downloaded and unpacked from the SRA using its accession ID and the SRA toolkit (specifically, prefetch and fasterqdump)

- the sample was mapped against the above indexed genome using STAR. If more than 50% of reads failed to map due to being "too short" then the sample was reprocessed using slightly relaxed mapping parameters (--outFilterScoreMinOverLread 0.3 --outFilterMatchNminOverLread 0.3). Regardless of how many reads were then mapped these reprocessed samples, and those that had mapped > 50% in the first run, were uploaded to CZID.org to run through their pipeline.

- once all had been completely through this pipeline, the data was downloaded locally for analysis. (input/CZID_TaxonReports/)

## Uploading the Data

The data exists in the input/CZID_TaxonReports/ directory. In order to properly label and process it, I also need the sample metadata.

```{r}
met <- read.table("input/SRA/MetaData_Edit_Oct22.tsv", 
                  sep = "\t", header = T)

head(met)
```

```{r}
csvs <- list.files(path = "input/CZID_TaxonReports/", pattern = "*.csv")
head(csvs, n = 20)
```

```{r}
data.raw <- vector(mode = "list", length = length(csvs))

for (i in 1:length(csvs)){
  data.raw[[i]] <- read.csv(paste("input/CZID_TaxonReports/", csvs[i], sep = ""),
                              header = T)
}

head(data.raw[[330]])
```
In order to properly filter the data, I need taxonomic information

## Making Taxonomy Files

### The Tax File

So I downloaded a taxonomy dump from NCBI on 18th October 2022 and I'm going to use that as my taxonomy resource. However, it is huge, too big for me to process it either in R or Sublime. So I've split into smaller tables which I will process through, removing anything I don't need, and then bringing them altogether at the end (likely in the commandline)

Edits that need to happen:

1. I only want these files to go to genus level so tax_id and species names can be removed. Similarly, any entries that don't have a genera attached can be taken away.

2. All vertebrates can be removed

3. All invertebrates can be removed. I can get mite/worm genealogy separately.

I also need a separate file for viral taxonomy that goes all the way down to species level as they don't have easy indicators of genus (ie Black queen cell virus = Triatovirus, Deformed wing virus = Iflavirus) and there are too many to try and make these annotations manually.

First I altered the .dmp file in a plain text editor so that the pipe deliminaters were replaced with tabs. I then ran this script.

After this is ran, all editing chunks will be set to eval = FALSE, as the required files will no longer exist.

```{bash, eval = FALSE}
#usr/bin/bash

split -l 50000 rankedlineage_181022.tsv rankedLinSplit_

for f in rankedLinSplit*; do                          
        mv "$f" "$f".tsv
done
```

Now for the r part of the processing.

First. Viruses.

```{r, eval = FALSE}
taxfiles <- list.files(path = "input/Phylo_Misc/", pattern = "*.tsv")
#remove the original file
taxfiles <- taxfiles[c(-1,-2,-52)]
taxfiles

#make space to store viral entries
#no idea how many there will be, so let's guess maybe 10? Safe side.
vir.list <- vector(mode = "list", length = 10)

j <- 1
for (i in 1:length(taxfiles)){
  x <- read.table(paste("input/Phylo_Misc/", taxfiles[i], sep = ""), 
                   comment.char = "", header = F, sep = "\t", quote = "")
  #take the header off the first file
  if(i == 1){
    x <- x[-1,]
  }
  x <- subset(x, V10 == "Viruses")
  v.check <- nrow(x)
  if (v.check > 1){
    print(j)
    vir.list[[j]] <- x
    j <- j + 1
  }
}

vir.df <- bind_rows(vir.list)

# remove unnecessary columns
vir.df <- vir.df %>% 
  mutate_all(na_if, "")

vir.df <- vir.df[,c(2:10)]

names(vir.df) <-  c("Common name", "Species", "Genus", "Family", "Order", "Class", "Phylum", "Kingdom", "SuperKingdom")

write.table(vir.df, "input/Phylo_Misc/ViralTaxo_Oct22.tsv",
            quote = F, row.names = F, col.names = T, sep = "\t")
```

And now for the rest of the taxonomy

```{r, eval = FALSE}
#define what's getting kept: microbial (bacteria/archaea/viruses/single celled)
wantedSuper <- c("Bacteria", "Archaea")
wantedKing <- c("Fungi", "Ciliophora", "Choanoflagellata", "Heterolobosea",
                "Evosea", "Apicomplexa", "Oomycota", "Bigyra")
wantedOrder <- c("Trypanosomatida", "Trombidiformes")
wantedPhylum <- c("Nematoda")
wantedGenus <- c("Apocephalus", "Aethina")

for (i in 1:length(taxfiles)){
  x <- read.table(paste("input/Phylo_Misc/", taxfiles[i], sep = ""), 
                   comment.char = "", header = F, sep = "\t", quote = "")
  if(i == 1){
    x <- x[-1,]
  }
  x <- x[,c(4:10)]
  #don't include lines that don't have a genus entry
  x <- x[!x$V4 == "",]
  x2 <- x[x$V10 %in% wantedSuper |
          x$V9 %in% wantedKing |
          x$V6 %in% wantedOrder |
          x$V4 %in% wantedGenus |
          x$V8 %in% wantedPhylum,]
  write.table(x2, paste("input/Phylo_Misc/", taxfiles[i], sep = ""),
              col.names = F, row.names = F, sep = "\t", quote = F)
}
```

Will cat those files together in the commandline

```{r}
tax <- read.table("input/Phylo_Misc/rankedlin_Edit_Oct22.tsv",
                  header = F, sep = "\t", quote = "", comment.char = "")
names(tax) <- c("Genus", "Family", "Order", "Class", "Phylum", "Kingdom", "SuperKingdom")
#remove duplicates
tax <- tax[!duplicated(tax),]

summary(as.factor(tax$SuperKingdom))
```

## Filtering data

I will filter the data in order to only keep hits that are more likely (NR and NT > 1 and percent_id > 50%). 

NB: sometimes things can exist that are only in NR or NT but not the other database. A taxon can contain high NT and low NR if rRNA is present in NT and not NR. The opposite can also be true: a divergent virus may match NR but be too divergent to hit within NT. This is a point that should be considered particularly in the case of viruses.

I will filter through twice. Once, keeping all non-viruses and applying the filter described above. The second, keeping all viruses and only applying the NR filter parameters. The two will be recombined at the end.

Further steps: 

- the non-virus taxa will have to be further filtered for reads not wanted, i.e. bee or hits within Hexapoda

I will then filter by species - only keeping eukaryotic hits that aren't bees/insects/whatever other strange hits there are.

```{r}
head(data.raw[[2]])
```

### Non-viruses

```{r}
data.filt.nv <- vector(mode = "list", length = length(data.raw))

for (i in 1:length(data.filt.nv)){
  #only keep necessary columns
  data.filt.nv[[i]] <- data.raw[[i]][,c(4,6,8,11,12,15,23,24,27)]
  #get sample ID from the filename
  samp <- paste(strsplit(csvs[i], "_2[0-9]")[[1]][1])
  samp <- str_remove(samp, "_[1-2]")
  #add sample ID to dataframe for ease of identification when these are combined.
  data.filt.nv[[i]]$SampleID <- paste(samp)
  #get species  and genus information from the metadata object
  data.filt.nv[[i]]$Species <- unique(met$Species[met$Sample.ID == samp])
  data.filt.nv[[i]]$Genus <- unique(met$Genus[met$Sample.ID == samp])
  #filter the results by the parameters outlined above
  data.filt.nv[[i]] <- subset(data.filt.nv[[i]], nt_rpm > 1 &
                            nt_percent_identity > 50 &
                            nr_rpm > 1 &
                            nr_percent_identity > 50)
}


head(data.filt.nv[[69]])
```

Add euk genera information for ease of removal

```{r}
filt.df.nv <- bind_rows(data.filt.nv)
#remove viruses
filt.df.nv <- subset(filt.df.nv, category != "viruses")

for(i in 1:nrow(filt.df.nv)){
  gen <- strsplit(filt.df.nv$name[i], " ")
  filt.df.nv$NonHostPhylo[i] <- paste(gen[[1]][1])
}

#for some reason the dataframe has amassed rows of all NAs. These are causing issues and need to be removed
filt.df.nv <- filt.df.nv[grepl("^NA", rownames(filt.df.nv))==F,]

#how to deal with the parts that don't have genera?
#the genus will instead be the Family name (these have been added manually to the euk
#filtering file)
for (i in 1:nrow(filt.df.nv)){
  if (filt.df.nv$NonHostPhylo[i] == "non-genus-specific"){
    filt.df.nv$NonHostPhylo[i] <- strsplit(filt.df.nv$name[i], " ")[[1]][5]
  }
}

#there's also this guy in there
filt.df.nv$NonHostPhylo[filt.df.nv$name == "parasitid 'Pas'"] <- "Parasitoidea"

summary(as.factor(filt.df.nv$NonHostPhylo))
```

Load up eukaryote classifications and remove those that are unwanted

```{r}
euk <- read.csv("input/Phylo_Misc/Eukaryota_Classifcations_Oct22.csv")

#make lists of unwanted eukaryota
tmp <- unique(euk$Species[grepl("Insect|Arachnid|Plant|Annelid|Crustacean|Myriapod|Aquatic",
                     euk$Classification)])
tmp[215] <- "Mushroom"
torem <- euk$Genus[euk$Species %in% tmp]

#remove whitespace
torem <- trimws(torem)

filt.df.nv <- filt.df.nv[!filt.df.nv$NonHostPhylo %in% torem,]

head(filt.df.nv)
```

I would also like to easily be able to pull out fungi

```{r}
fungi <- euk$Genus[euk$Classification == "Fungus"]

#add fungus to classifications
filt.df.nv$category[filt.df.nv$NonHostPhylo %in% fungi] <- "fungus"

#change eukaryota classification to make more sense
filt.df.nv$category[filt.df.nv$category == "eukaryota"] <- "other_eukaryota"

summary(as.factor(filt.df.nv$category))
```


### Viruses

```{r}
data.filt.v <- vector(mode = "list", length = length(data.raw))

for (i in 1:length(data.filt.v)){
  #only keep necessary columns
  data.filt.v[[i]] <- data.raw[[i]][,c(4,6,8,11,12,15,23,24,27)]
  #get sample ID from the filename
  samp <- paste(strsplit(csvs[i], "_2[0-9]")[[1]][1])
  samp <- str_remove(samp, "_[1-2]")
  #add sample ID to dataframe for ease of identification when these are combined.
  data.filt.v[[i]]$SampleID <- paste(samp)
  #get species  and genus information from the metadata object
  data.filt.v[[i]]$Species <- unique(met$Species[met$Sample.ID == samp])
  data.filt.v[[i]]$Genus <- unique(met$Genus[met$Sample.ID == samp])
  #filter the results by the parameters outlined above
  data.filt.v[[i]] <- subset(data.filt.v[[i]], nr_rpm > 1 &
                            nr_percent_identity > 50)
}

filt.df.v <- bind_rows(data.filt.v)
filt.df.v <- subset(filt.df.v, category == "viruses")

head(filt.df.v)
```

Adding NonHostPhylo information will be more difficult for the viruses.

Genus is not wholly resolved in many of the species, nor, somehow, is Family/Order, so maybe going up to Class level would make more sense.

Also, some of the names form CZID are out of date. I found the updated names of the viruses failing to get a hit using the International Committee on Taxonomy of Viruses webpage (Citation of the ICTV Database using our manuscript in Nucleic Acids Research: "Virus taxonomy: the database of the International Committee on Taxonomy of Viruses (ICTV)". Lefkowitz et al Nucleic Acids Res. 2017 Oct 13. PubMed PMID: 29040670.For the most recent taxonomy release, you can use the following Archives of Virology manuscript: Walker et al Recent changes to virus taxonomy ratified by the International Committee on Taxonomy of Viruses (2022). Arch Virol. 2022 Aug 23. doi: 10.1007/s00705-022-05516-5. Epub ahead of print. PMID: 35999326.)

```{r}
vir.up <- read.table("input/Phylo_Misc/ViralUpdate.txt",
                     header = T, sep = "\t")

#now to update the old names with the new 
for (i in 1:nrow(filt.df.v)){
  if(filt.df.v$name[i] %in% vir.up$Old_name){
    filt.df.v$name[i] <- vir.up$New_name[vir.up$Old_name == filt.df.v$name[i]]
  }
}
```

And now to classify....

For those that didn't exist in the ICTV, I added entries manually using the NCBI taxonomy service (Schoch CL, et al. NCBI Taxonomy: a comprehensive update on curation, resources and tools. Database (Oxford). 2020: baaa062. PubMed: 32761142 PMC: PMC7408187)

```{r}
#read in viral taxonomy
vir.df <- read.table("input/Phylo_Misc/ViralTaxo_Oct22.tsv", 
                      header = T, sep = "\t", quote = "", comment.char = "")

#run through the viruses and, if possible, extract Class information from the above file
#any hits that have an NA return either don't have Class information (some are only
#classified to Class level, etc) or there are multiple entries or where the the taxa
#hit is already a genus
filt.df.v$NonHostPhylo <- paste("")
for (i in 1:nrow(filt.df.v)){
  if (filt.df.v$name[i] %in% vir.df$Common.name){
    t <- unique(vir.df$Class[vir.df$Common.name == filt.df.v$name[i]])
    #remove any NAs
    if (length(t) > 1){
     t <- t[!is.na(t)] 
    }
    filt.df.v$NonHostPhylo[i] <- paste(t)
  }
}

#control for "non-genus-specific" 
for (i in 1:nrow(filt.df.v)){
  if(grepl("non-genus-specific", filt.df.v$name[i]) == T){
    x <- strsplit(filt.df.v$name[i], " ")[[1]][5]
    filt.df.v$NonHostPhylo[i] <- paste(unique(vir.df$Class[vir.df$Common.name == x]))
    
  }
}

#all that's left from here is either viruses listed in the taxa file that are only 
#classified to the virus level or viruses that don't exist in the taxa file
#so either way ... unclassified viruses
filt.df.v$NonHostPhylo[filt.df.v$NonHostPhylo == "NA"] <- "Unclassified virus"
```

###Combine

And now ... finally ... I can combine

```{r}
filt.df <- rbind(filt.df.nv, filt.df.v)
```

And lastly, remove an entry that causes issues later on and is generally just very uninformative.

```{r}
filt.df <- subset(filt.df, 
                  name != "all taxa with neither family nor genus classification")
```


## Making a PhyloSeq Object

Specifically using: https://joey711.github.io/phyloseq/import-data.html#loading_included_data

Phyloseq objects are used by the majority of packages in r that do things such as estimate species richness for microbial datasets. In essence, I need to provide a number of dataframes in certain formats to combine into a phyloseq object - mainly, a counts table and a taxonomy classifications table (with sample metadata).

### Counts table

There are two count tables I can produce: 1) in nr_rPM or 2) nr_counts. The former is normalised which may be an issue with the packages as the other input I'm seeing are straight up counts tables that have not been transformed in any way. The counts are the raw number of read hits and probably the most similar to the OTU tables that these methods usually use as input.

Requirements: numerical matrix, rows = microbials, columns = samples.

Running this brought to light a couple of duplicates per sample. When this occurred I removed the duplicates manually, keeping whichever had the largest percent identity. 

```{r}
rownames(filt.df) <- 1:nrow(filt.df)
filt.df[filt.df$SampleID == "SRR13769017" & filt.df$name == "Candida", ]
filt.df[19625,]

filt.df <- filt.df[-9132,] # SRR12527929 had 2 Candida hits
rownames(filt.df) <- 1:nrow(filt.df)#reset

filt.df <- filt.df[-19010,] # SRR13769016 had 2 Candida hits
rownames(filt.df) <- 1:nrow(filt.df) 

filt.df <- filt.df[-19625,] # SRR13769017 had 2 Candida hits
rownames(filt.df) <- 1:nrow(filt.df) 

```


```{r}
#get a list of all the microbials that have been detected across all samples
micro <- unique(filt.df$name)

#compile the samples
samples <- unique(filt.df$SampleID)

cnt <- matrix(nrow = length(micro), ncol = length(samples))
for (i in 1:length(samples)){
  print(samples[i])
  x <- vector(length = length(micro))
  for (j in 1:length(x)){
    t <- filt.df$nr_count[filt.df$name == micro[j] &
                        filt.df$SampleID == samples[i]]
    if (length(t)==0){
    t <- 0
    }
    x[j] <- t
  }
  cnt[,i] <- x
}
```

That took an impressively long time so gonna save before continuing

```{r}
rownames(cnt) <- micro
colnames(cnt) <- samples

dir.create("output/Counts/")
write.table(cnt, "output/Counts/nr_Counts_raw.tsv",
            row.names = T, col.names = T, sep = "\t", quote = F)
```













