---
title: "Anthophila_Microbial_Comparison_Proj"
author: "Lauren `Mee"
date: "2022-09-27"
output: html_document
---

27th September 2022

```{r}
set.seed(1517)
dir.create("output/")
```

## Libraries

```{r}
library(ggplot2)
library(dplyr)
library(stringr)
library(reshape)
library(phyloseq)
```

## Resources

Tutorial discussing different methods and approaches to assesing microbiome data from metagenomic sources:
https://yanhui09.github.io/microbiome_analysis/1_microbiome_r.html#what-you-will-work-on
(includes DESeq2 approaches and a good heatmap alternative that looks interesting)

Microbial Community Diversity Analysis Tutorial with Phyloseq:
http://deneflab.github.io/MicrobeMiseq/demos/mothur_2_phyloseq.html
(may not be as useful, looks to be explicitly for amplicon data)

Introduction to the Statistical Analysis of Microbiome Data in R
https://www.nicholas-ollberding.com/post/introduction-to-the-statistical-analysis-of-microbiome-data-in-r/

Tutorial for making phyloseq objects
https://joey711.github.io/phyloseq/import-data.html

"Breakaway" Approach to estimating richness
https://adw96.github.io/breakaway/articles/intro-diversity-estimation.html


## Functions






## Experiment Notes

The purpose of this project is to data mine existing publc RNAseq and other resources and use the CZID.org tool in order to elucidate potential patterns of microbial content across the sample species used therein. 

Samples were chosen from SRA projects that used Hymenopteran species and also met the following criteria: 1) the tissue used was or included the abdomen, 2) nothing experimental had happened to the individual sample that would be an obvious influence upon the microbiome, i.e. no infection administered / particularly stressed individuals and 3) sociality of the sample was known. 

These were then processed in batches using the script ProcessSRA_v3.sh (Scripts/Bash/). Lists of samples were arranged in batch files with one sample accession id per line. The script then would process through these lists. Each set of individual sample reads were downloaded and unpacked from the NCBI SRA using their SRA toolkit, before being uploaded to CZID.org using the CZID command line interface. CZID.org then began its own pipeline, sorting through reads that do not match the "host" species (a generic "bee" option that is actually Apis mellifera) and mapping them against other species. This can then be used to interpret the potential microbial content of the sample as reads match to viruses / bacteria / eukaryotic parasites.

## Sample Metadata

Before I can begin any analyses, I need to know what I'm working with.

```{r}
met <- read.table("input/SRA/SampleMetadata.tsv", sep = "\t", header = T)

head(met)
```

I don't really need most of these fields. I want to reduce collection date to just years and abstracts etc aren't necessary at all

```{r}
for (i in 1:nrow(met)){
  end <- strsplit(met$Collection.Date[i], "-")[[1]][2]
  met$YearCollected[i] <- paste(20, end, sep = "")
}

met$Abstract <- NULL
met$Collection.Date <- NULL
met$Publication <- NULL

names(met)[7] <- "NucleotideType"

head(met)
```

Now to get some overall numbers.

```{r}
nrow(met)
```

How many, and how many of each, species ?

```{r}
length(unique(met$Species))
summary(as.factor(met$Species))
```

Hmm. I'm probably gonna have reduce by genus / family here.

```{r}
for (i in 1:nrow(met)){
  genus <- strsplit(met$Species[i], " ")[[1]][1]
  met$Genus[i] <- paste(genus)
}

head(met)
```
```{r}
length(unique(met$Genus))
summary(as.factor(met$Genus))
```

Adding Family and Tribe

```{r}
specs <- c("Anthophora", "Apis", "Bombus", "Ceratina", "Epeolus", 
            "Eufriesea", "Euglossa", "Exoneura", "Habropoda", "Nomada",
            "Tetragonisca", "Tetragonula", "Thyreus", "Xylocopa")
fams <- rep("Apidae", length(specs))
subfams <- c(rep("Apinae", 3), "Xylocopinae", "Nomadinae", 
             rep("Apinae", 2), "Xylocopinae", "Apinae", "Nomadinae",
             rep("Apinae", 3), "Xylocopinae")
tribes <- c("Anthophorini", "Apini", "Bombini", "Ceratini", "Epeolini", 
            "Euglossini", "Euglossini", "Allodapini", "Anthophorini", "Nomadini",
            "Meliponini", "Meliponini", "Melectini", "Xylocopini")

specs[15:18] <- c("Dufourea", "Halictus", "Lasioglossum", "Megalopta")
fams[15:18] <- "Halictidae"
subfams[15:18] <- c("Rophitinae", rep("Halictinae", 3))
tribes[15:18] <- c("Rophitini", "Halictini", "Halictini", "Augochlorini")

specs[19] <- c("Andrena")
fams[19]  <- "Andrenidae"
subfams[19] <- "Andreninae"
tribes[19] <- "Andrenini"

specs[20] <- "Colletes"
fams[20] <- "Colletidae"
subfams[20] <- "Colletinae"
tribes[20] <- "Colletini"
  
specs[21] <- "Osmia"
fams[21] <- "Megachilidae"
subfams[21] <- "Megachilinae"
tribes[21] <- "Osmiini"

key <- as.data.frame(cbind(specs, fams, subfams, tribes))
key
```

```{r}
for (i in 1:nrow(met)){
  met$Family[i] <- key$fams[key$specs == paste(met$Genus[i])]
  met$SubFamily[i] <- key$subfams[key$specs == paste(met$Genus[i])]
  met$Tribe[i] <- key$tribes[key$specs == paste(met$Genus[i])]
}
met
```
Clean up

```{r}
remove(fams, specs, subfams, tribes)
```

Sociality

```{r}
met$Soc <- met$Sociality
met$Sociality[grep("olymorphic",met$Soc)] <- "Polymorphic"
met$Sociality[grep("Solitary",met$Soc)] <- "Solitary"
met$Sociality[met$Soc == "Solitary/Polymorphic?"] <- "Polymorphic"
met$Soc <- NULL
```


```{r}
table(met$Sociality, met$NucleotideType)
```

```{r}
head(met)
```
Save

```{r}
#write.table(met, "output/MetaData_Edit_Oct22.tsv",
 #           sep = "\t", row.names = F, col.names = T, quote = F)
```

## Making Taxonomy Files

So I downloaded a taxonomy dump from NCBI on 18th October 2022 and I'm going to use that as my taxonomy resource. However, it is huge, too big for me to process it either in R or Sublime. So I've split into smaller tables which I will process through, removing anything I don't need, and then bringing them altogether at the end (likely in the commandline)

Edits that need to happen:

1. I only want these files to go to genus level so tax_id and species names can be removed. Similarly, any entries that don't have a genera attached can be taken away.

2. All vertebrates can be removed

3. All invertebrates can be removed. I can get mite/worm genealogy separately.

I also need a separate file for viral taxonomy that goes all the way down to species level as they don't have easy indicators of genus (ie Black queen cell virus = Triatovirus, Deformed wing virus = Iflavirus) and there are too many to try and make these annotations manually.

First I altered the .dmp file in a plain text editor so that the pipe deliminaters were replaced with tabs. 

All editing chunks will be set to eval = FALSE, as the required files will no longer exist.

```{bash, eval = FALSE}
#usr/bin/bash

split -l 50000 rankedlineage_181022.tsv rankedLinSplit_

for f in rankedLinSplit*; do                          
        mv "$f" "$f".tsv
done
```

Now for the r part of the processing.

First. Viruses.

```{r, eval = FALSE}
taxfiles <- list.files(path = "input/Phylo_Misc/", pattern = "*.tsv")
#remove the original file
taxfiles <- taxfiles[c(-1,-2)]
taxfiles

#make space to store viral entries
#no idea how many there will be, so let's guess maybe 10? Safe side.
vir.list <- vector(mode = "list", length = 10)

j <- 1
for (i in 1:length(taxfiles)){
  x <- read.table(paste("input/Phylo_Misc/", taxfiles[i], sep = ""), 
                   comment.char = "", header = F, sep = "\t", quote = "")
  #take the header off the first file
  if(i == 1){
    x <- x[-1,]
  }
  x <- subset(x, V10 == "Viruses")
  v.check <- nrow(x)
  if (v.check > 1){
    print(j)
    vir.list[[j]] <- x
    j <- j + 1
  }
}

vir.df <- bind_rows(vir.list)

#I don't want lines that have broken information. Only those that go from name all the way through to superkingdom.
#replace blanks with NA and then remove any lines with NA
#then remove unnecessary columns
vir.df <- vir.df %>% 
  mutate_all(na_if, "")

vir.df <- vir.df[complete.cases(vir.df),]

vir.df <- vir.df[,c(2:10)]

names(vir.df) <-  c("Common name", "Species", "Genus", "Family", "Order", "Class", "Phylum", "Kingdom", "SuperKingdom")

write.table(vir.df, "input/Phylo_Misc/ViralTaxo_Oct22.tsv",
            quote = F, row.names = F, col.names = T, sep = "\t")
```

And now for the rest of the taxonomy

```{r, eval = FALSE}
#define what's getting kept: microbial (bacteria/archaea/viruses/single celled)
wantedSuper <- c("Bacteria", "Archaea")
wantedKing <- c("Fungi", "Ciliophora", "Choanoflagellata", "Heterolobosea",
                "Evosea", "Apicomplexa", "Oomycota", "Bigyra")
wantedOrder <- c("Trypanosomatida", "Trombidiformes")
wantedPhylum <- c("Nematoda")
wantedGenus <- c("Apocephalus", "Aethina")

for (i in 1:length(taxfiles)){
  x <- read.table(paste("input/Phylo_Misc/", taxfiles[i], sep = ""), 
                   comment.char = "", header = F, sep = "\t", quote = "")
  if(i == 1){
    x <- x[-1,]
  }
  x <- x[,c(4:10)]
  #don't include lines that don't have a genus entry
  x <- x[!x$V4 == "",]
  x2 <- x[x$V10 %in% wantedSuper |
          x$V9 %in% wantedKing |
          x$V6 %in% wantedOrder |
          x$V4 %in% wantedGenus |
          x$V8 %in% wantedPhylum,]
  write.table(x2, paste("input/Phylo_Misc/", taxfiles[i], sep = ""),
              col.names = F, row.names = F, sep = "\t", quote = F)
}
```

Will cat those files together in the commandline

```{r}
tax <- read.table("input/Phylo_Misc/rankedlin_Edit_Oct22.tsv",
                  header = F, sep = "\t", quote = "", comment.char = "")
names(tax) <- c("Genus", "Family", "Order", "Class", "Phylum", "Kingdom", "SuperKingdom")
#remove duplicates
tax <- tax[!duplicated(tax),]

summary(as.factor(tax$SuperKingdom))
```

## Uploading the Data

There's an individual taxon results file per sample (therefore 325 files, plus the duplicates) that I'll need to massage to make more amenable to my purposes. The final goal would be making a combined data table with some sort of measurement per species. It may be that for the combined tables I'll only be able to do so for one metric at a time.

NB: I do not yet have a results file per entry in the metadata (the extra Caus samples) so I'll use the old met file for now.

```{r}
met <- read.table("output/MetaData_Edit_Sep22.tsv", 
                  sep = "\t", header = T)
```


```{r}
csvs <- list.files(path = "input/CZID_TaxonReports/", pattern = "*.csv")
head(csvs, n = 20)
```

```{r}
data.raw <- vector(mode = "list", length = length(csvs))

for (i in 1:length(csvs)){
  data.raw[[i]] <- read.csv(paste("input/CZID_TaxonReports/", csvs[i], sep = ""),
                              header = T)
}

head(data.raw[[330]])
```


## Filtering data

I will filter the data in order to only keep hits that are more likely (NR and NT > 1 and percent_id > 50%). 

NB: sometimes things can exist that are only in NR or NT but not the other database. A taxon can contain high NT and low NR if rRNA is present in NT and not NR. The opposite can also be true: a divergent virus may match NR but be too divergent to hit within NT. This is a point that should be considered particularly in the case of viruses.

I will filter through twice. Once, keeping all non-viruses and applying the filter described above. The second, keeping all viruses and only applying the NR filter parameters. The two will be recombined at the end.

Further steps: 

- the non-virus taxa will have to be further filtered for reads not wanted, i.e. bee or hits within Hexapoda

I will then filter by species - only keeping eukaryotic hits that aren't bees/insects/whatever other strange hits there are.

```{r}
head(data.raw[[2]])
```

### Non-viruses

```{r}
data.filt.nv <- vector(mode = "list", length = length(data.raw))

for (i in 1:length(data.filt.nv)){
  #only keep necessary columns
  data.filt.nv[[i]] <- data.raw[[i]][,c(4,6,8,11,12,15,23,24,27)]
  #get sample ID from the filename
  samp <- paste(strsplit(csvs[i], "_2[0-9]")[[1]][1])
  samp <- str_remove(samp, "_[1-2]")
  #add sample ID to dataframe for ease of identification when these are combined.
  data.filt.nv[[i]]$SampleID <- paste(samp)
  #get species  and genus information from the metadata object
  data.filt.nv[[i]]$Species <- unique(met$Species[met$Sample.ID == samp])
  data.filt.nv[[i]]$Genus <- unique(met$Genus[met$Sample.ID == samp])
  #filter the results by the parameters outlined above
  data.filt.nv[[i]] <- subset(data.filt.nv[[i]], nt_rpm > 1 &
                            nt_percent_identity > 50 &
                            nr_rpm > 1 &
                            nr_percent_identity > 50)
}


head(data.filt.nv[[69]])
```

Add euk genera information for ease of removal

```{r}
filt.df.nv <- bind_rows(data.filt.nv)
#remove viruses
filt.df.nv <- subset(filt.df.nv, category != "viruses")

for(i in 1:nrow(filt.df.nv)){
  gen <- strsplit(filt.df.nv$name[i], " ")
  filt.df.nv$NonHostGenus[i] <- paste(gen[[1]][1])
}

#for some reason the dataframe has amassed rows of all NAs. These are causing issues and need to be removed
filt.df.nv <- filt.df.nv[grepl("^NA", rownames(filt.df.nv))==F,]

#how to deal with the parts that don't have genera?
#the genus will instead be the Family name (these have been added manually to the euk
#filtering file)
for (i in 1:nrow(filt.df.nv)){
  if (filt.df.nv$NonHostGenus[i] == "non-genus-specific"){
    filt.df.nv$NonHostGenus[i] <- strsplit(filt.df.nv$name[i], " ")[[1]][5]
  }
}

#there's also this guy in there
filt.df.nv$NonHostGenus[filt.df.nv$name == "parasitid 'Pas'"] <- "Parasitoidea"

summary(as.factor(filt.df.nv$NonHostGenus))
```

Load up eukaryote classifications and remove those that are unwanted

```{r}
euk <- read.csv("input/Phylo_Misc/Eukaryota_Classifcations_Oct22.csv")

#make lists of unwanted eukaryota
tmp <- unique(euk$Species[grepl("Insect|Arachnid|Plant|Annelid|Crustacean|Myriapod|Aquatic",
                     euk$Classification)])
tmp[215] <- "Mushroom"
torem <- euk$Genus[euk$Species %in% tmp]

#remove whitespace
torem <- trimws(torem)

filt.df.nv <- filt.df.nv[!filt.df.nv$NonHostGenus %in% torem,]

head(filt.df.nv)
```
### Viruses

```{r}
data.filt.v <- vector(mode = "list", length = length(data.raw))

for (i in 1:length(data.filt.v)){
  #only keep necessary columns
  data.filt.v[[i]] <- data.raw[[i]][,c(4,6,8,11,12,15,23,24,27)]
  #get sample ID from the filename
  samp <- paste(strsplit(csvs[i], "_2[0-9]")[[1]][1])
  samp <- str_remove(samp, "_[1-2]")
  #add sample ID to dataframe for ease of identification when these are combined.
  data.filt.v[[i]]$SampleID <- paste(samp)
  #get species  and genus information from the metadata object
  data.filt.v[[i]]$Species <- unique(met$Species[met$Sample.ID == samp])
  data.filt.v[[i]]$Genus <- unique(met$Genus[met$Sample.ID == samp])
  #filter the results by the parameters outlined above
  data.filt.v[[i]] <- subset(data.filt.v[[i]], nr_rpm > 1 &
                            nr_percent_identity > 50)
}

head(data.filt.v[[2]])
```

```{r}
filt.df.v <- bind_rows(data.filt.v)
filt.df.v <- subset(filt.df.v, category == "viruses")

head(filt.df.v)
```
Adding NonHostGenus information will be more difficult for the viruses.

```{r}
#read in viral taxonomy
vir.tax <- read.table("input/Phylo_Misc/ViralTaxo_Oct22.tsv", 
                      header = T, sep = "\t", quote = "", comment.char = "")

#run through the viruses and, if possible, extract genus information from the above file
#first, using the common name, then by species, finally by genus
filt.df.v$NonHostGenus <- paste("")
for (i in 1:nrow(filt.df.v)){
  if (filt.df.v$name[i] %in% vir.tax$Common.name){
    filt.df.v$NonHostGenus[i] <- unique(vir.tax$Genus[vir.tax$Common.name ==
                                                        filt.df.v$name[i]])
  }
  if (filt.df.v$name[i] %in% vir.tax$Species){
    filt.df.v$NonHostGenus[i] <- unique(vir.tax$Genus[vir.tax$Species ==
                                                        filt.df.v$name[i]])
  }
  if (filt.df.v$name[i] %in% vir.tax$Genus){
    filt.df.v$NonHostGenus[i] <- unique(vir.tax$Genus[vir.tax$Genus ==
                                                        filt.df.v$name[i]])
  }
}

```

That didn't work for everything. What's left?

```{r}
unique(filt.df.v$name[filt.df.v$NonHostGenus == ""])
```

```{r}
vir.tax[vir.tax$Species == "Varroa destructor virus 1",]
```


## Making a PhyloSeq Object

To start with, I need the equivalent of an OTU table. This is essentially a counts table with rownames being taxon detected by CZID and colnames being each sample, the matrix itself consisting either of reads or rPM.

Because of the viruses sometimes hitting higher in NR databases than NT and because I don't want to discount these (where NR hits are high but NR are 0 or very low), I'll be making two tables: one with NR_rPM and one with NR_reads.


```{r}
head(filt.df)
```

```{r}
head(data.raw[[2]])
```




## Analysing Data

Let's run through everything with this data anyway, in preparation for my "final" data, should I ever receive it

What are my questions ? 

- Are there any patterns among microbial composition and sociality? Mainly, is solitary microbiota more diverse than social, as would be expected due to lack of strong horizontal transfer? Or are these patterns, should they exist, more associated with phylogenetic relationships? 

- Are there non-host genera in common between different bees? Is this genus- or tribe- or sociality- specific?

- Are there strong correlating relationships between certain host and non-host genera?

There are some possible other avenues, ie looking at plant-host associations, or sources of contamination (where have all these aquatic hits came from?) but we have a timeline here people.

